h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -678,114 +697,311 @@ public DataType getKeyType() {
-        private X current;
 
-        TMIterator(TransactionMap<K,?> transactionMap, K from, K to, boolean includeAllUncommitted) {
+        private final BitSet snapshotCommittingTransactions;
+
+        private final Cursor<K, VersionedValue> snapshotCursor;
+
+        private K snapshotKey;
+
+        private Object snapshotValue;
+
+        private final Cursor<K, VersionedValue> uncommittedCursor;
+
+        private K uncommittedKey;
+
+        private Object uncommittedValue;
+
+        RepeatableIterator(TransactionMap<K, ?> transactionMap, K from, K to, boolean forEntries) {
+            super(forEntries);
+            keyType = transactionMap.map.getKeyType();
             Transaction transaction = transactionMap.getTransaction();
             this.transactionId = transaction.transactionId;
-            TransactionStore store = transaction.store;
-            MVMap<K, VersionedValue> map = transactionMap.map;
             // The purpose of the following loop is to get a coherent picture
             // of a state of two independent volatile / atomic variables,
             // which they had at some recent moment in time.
             // In order to get such a "snapshot", we wait for a moment of silence,
             // when neither of the variables concurrently changes it's value.
             BitSet committingTransactions;
             RootReference mapRootReference;
-            if (includeAllUncommitted) {
-                do {
-                    committingTransactions = store.committingTransactions.get();
-                    mapRootReference = map.flushAndGetRoot();
-                } while (committingTransactions != store.committingTransactions.get());
-            } else {
-                do {
-                    committingTransactions = transactionMap.getCommittingTransactions();
-                    mapRootReference = transactionMap.getRootReference();
-                } while (committingTransactions != transactionMap.getCommittingTransactions());
-            }
+            do {
+                committingTransactions = transactionMap.getCommittingTransactions();
+                mapRootReference = transactionMap.getRootReference();
+            } while (committingTransactions != transactionMap.getCommittingTransactions());
             // Now we have a snapshot, where mapRootReference points to state of the map
             // and committingTransactions mask tells us which of seemingly uncommitted changes
             // should be considered as committed.
             // Subsequent map traversal uses this snapshot info only.
-            this.cursor = new Cursor<>(mapRootReference.root, from, to);
-            this.committingTransactions = committingTransactions;
+            this.snapshotCursor = new Cursor<>(mapRootReference.root, from, to);
+            this.snapshotCommittingTransactions = committingTransactions;
+
+            this.uncommittedCursor = new Cursor<>(transactionMap.map.flushAndGetRoot().root, from, to);
```

Comment:
This is not a snapshot at the beginning of current command, but just current snapshot, which is not the same. It will see all the changes made by this transaction since current command started, but it should not.

Reply:
It could be acceptable, but I'm writing a change for it right now to collect all of them for each command.